.github/workflows/build-and-test.yml

```
@@ -0,0 +1,94 @@
+name: build and test
+
+defaults:
+  run:
+    shell: bash -eo pipefail {0}
+
+on:
+  workflow_call:
+    inputs:
+      cuda_version: {type: string, required: true}
+      compiler: {type: string, required: true}
+      compiler_exe: {type: string, required: true}
+      compiler_version: {type: string, required: true}
+      std: {type: string, required: true}
+      gpu_build_archs: {type: string, required: true}
+      cpu: {type: string, required: true}
+      os: {type: string, required: true}
+      build_script: {type: string, required: false}
+      test_script: {type: string, required: false }
+
+jobs:
+  build:
+    if: inputs.build_script != ''
+    name: Build ${{inputs.compiler}}${{inputs.compiler_version}}/C++${{inputs.std}}
+    runs-on: ${{ (inputs.os == 'windows' && 'windows-latest') || format('linux-{0}-cpu16', inputs.cpu) }}
+    container: 
+      options: -u root
+      image: rapidsai/devcontainers:23.06-cpp-${{inputs.compiler}}${{inputs.compiler_version}}-cuda${{inputs.cuda_version}}-${{inputs.os}}
+    permissions:
+      id-token: write
+      contents: read
+    steps:
+      - name: Echo git version
+        run: git --version
+      #TODO: We can remove this when we aren't using submodules anymore
+      - name: Update git version
+        run: |
+          sudo apt-get install software-properties-common
+          sudo add-apt-repository ppa:git-core/ppa -y
+          sudo apt-get update
+          sudo apt-get install git -y
+          git --version
+      - name: Checkout repo
+        uses: actions/checkout@v3
+        with:
+          submodules: 'recursive'
+      - name: Configure credentials and environment variables for sccache
+        uses: ./.github/actions/configure_cccl_sccache
+      - name: Run build script
+        run: | 
+            sccache -s > sccache_before.txt
+            time ./${{ inputs.build_script }} "${{inputs.compiler_exe}}" "${{inputs.std}}" "${{inputs.gpu_build_archs}}"
+            sccache -s > sccache_after.txt
+      - name: Calculate sccache hit rate
+        run: |
+          hit_rate=$(./ci/sccache_hit_rate.sh sccache_before.txt sccache_after.txt  2>&1 >/dev/null)
+          echo "sccache hit rate: $hit_rate" >> $GITHUB_STEP_SUMMARY
+  test:
+    needs: build
+    if:  ${{ !cancelled() && ( needs.build.result == 'success' || needs.build.result == 'skipped' ) && inputs.test_script != '' }}
+    name: Test ${{inputs.compiler}}${{inputs.compiler_version}}/C++${{inputs.std}}
+    runs-on: linux-${{inputs.cpu}}-gpu-v100-latest-1
+    container: 
+      options: -u root
+      image: rapidsai/devcontainers:23.06-cpp-${{inputs.compiler}}${{inputs.compiler_version}}-cuda${{inputs.cuda_version}}-${{inputs.os}}
+      env:
+        NVIDIA_VISIBLE_DEVICES: ${{ env.NVIDIA_VISIBLE_DEVICES }}
+    permissions:
+      id-token: write
+      contents: read
```

```suggestion
      id-token: write
```

Same thing about job permissions here.